test/integration/interactive_test.js

```diff
@@ -0,0 +1,149 @@
+  });
+
+  teardown(async () => {
+    if (worker) {
+      await worker.terminate();
+      worker = null;
+    }
+  });
+
+  async function getWithoutRedirect (url) {
+    var res = await request.get(url).redirects(0).end();
+    return res.headers.location;
+  };
+
+  test('cat', async () => {
+    let taskId = slugid.v4();
+  	let task = {
+      payload: {
+        image: 'busybox',
+        command: cmd('sleep 15'),
+        maxRunTime: 2 * 60,
+        features: {
+          interactive: true
+        }
+      }
+    };
+    debug('posting to queue');
+    worker.postToQueue(task, taskId);
+    
+    var passed = false;
+
+    var signedUrl = worker.queue.buildSignedUrl(
+        worker.queue.getLatestArtifact,
+        taskId,
+        'private/mozilla/interactive.sock',
+        {expiration: 60 * 5});
+
+    var url;
+    await base.testing.poll(async () => {
+      url = await getWithoutRedirect(signedUrl);
+      assert(url, 'artifact not found');
+    }, 20, 1000);
+
+    //for testing, we don't care about https verification
+    var client = new DockerExecClient({
+      tty: false,
+      command: ['cat'],
+      url: url,
+      wsopts: {rejectUnauthorized: false}
+    });
+    await client.execute();
+
+    var buf = new Buffer([0xfa, 0xff, 0x0a]);
+    client.stdin.write(buf);
+    //message is small enough that it should be returned in one chunk
+    client.stdout.on('data', (message) => {
+      assert(buf[0] === message[0], 'message wrong!');
+      assert(buf[1] === message[1], 'message wrong!');
+      assert(buf[2] === message[2], 'message wrong!');
+      passed = true;
+      debug('test finished!');
+      client.close();
+    });
+
+    await new Promise(accept => client.socket.once('close', accept));
+    assert(passed);
+  });
+
+  test('cat stress test', async () => {
+    let taskId = slugid.v4();
+    let task = {
+      payload: {
+        image: 'busybox',
+        command: cmd('sleep 60'),
+        maxRunTime: 2 * 60,
+        features: {
+          interactive: true
+        }
+      }
+    };
+    debug('posting to queue');
+    worker.postToQueue(task, taskId);
+    
+    var passed = false;
+
+    var signedUrl = worker.queue.buildSignedUrl(
+        worker.queue.getLatestArtifact,
+        taskId,
+        'private/mozilla/interactive.sock',
+        {expiration: 60 * 5});
+
+    var url;
+    await base.testing.poll(async () => {
+      url = await getWithoutRedirect(signedUrl);
+      assert(url, 'artifact not found');
+    }, 20, 1000);
+
+    //for testing, we don't care about https verification
+    var client = new DockerExecClient({
+      tty: false,
+      command: ['cat'],
+      url: url,
+      wsopts: {rejectUnauthorized: false}
+    });
+    await client.execute();
+
+    const TEST_BUF_SIZE = 1024 * 1024;
+
+    var buf = await Promise.denodeify(crypto.pseudoRandomBytes)(TEST_BUF_SIZE);
+    // make sure the last character is a newline
+    buf[TEST_BUF_SIZE - 1] = 0x0a;
```

Comment:
curious, is that really necessary? Won't cat print if you signal end to input...
Oh, we probably need to have a message for signaling `end` of input... 

Basically if someone does `client.stdin.end()` they're saying that this is the end of the stream and it should close. Do we even send that right now?

If not then isn't the `cat` process killed by the docker container coming to an end.. .rather than terminating because the input stream is at an end... So we don't get exit code zero... And the cat subprocess remains alive forever.

Note, to close bash/cat/grep we just end stdin... This is how `cat myfile | grep needle` works... grep terminates because the stream given by stdin comes to an end. We need the ability to close the stream, note that this is different from aborting or killing the process... Because the process could still be writing back to us for a while before it stops... or it could choose not to stop or even care about the stdin...

To do real kill we probably need some tricks and smart utilities, but let's do that later...